tests/profile/test_accuracy.py

```diff
@@ -0,0 +1,100 @@
+# -*- encoding: utf-8 -*-
+import collections
+import time
+
+from ddtrace.profile import _compat
+from ddtrace.profile import profiler
+from ddtrace.profile.collector import stack
+
+
+def spend_1():
+    time.sleep(1)
+
+
+def spend_3():
+    time.sleep(3)
+
+
+def spend_4():
+    spend_3()
+    spend_1()
+
+
+def spend_7():
+    spend_3()
+    spend_1()
+    spend_cpu_3()
+
+
+def spend_16():
+    spend_4()
+    spend_7()
+    spend_cpu_2()
+    spend_3()
+
+
+def spend_cpu_2():
+    now = _compat.monotonic_ns()
+    # Active wait for 2 seconds
+    while _compat.monotonic_ns() - now < 2e9:
+        pass
+
+
+def spend_cpu_3():
+    # Active wait for 3 seconds
+    now = _compat.monotonic_ns()
+    while _compat.monotonic_ns() - now < 3e9:
+        pass
+
+
+# We allow 2% error:
+# The profiler might not be precise, but time.sleep is not either.
+TOLERANCE = 0.02
+# Use 5% accuracy for CPU usage, it's way less precise.
+CPU_TOLERANCE = 0.05
+
+
+def almost_equal(value, target, tolerance=TOLERANCE):
+    return target * (1 + tolerance) >= value >= target * (1 - tolerance)
+
+
+def total_time(time_data, funcname):
+    return sum(functime[funcname] for functime in time_data.values())
+
+
+def test_accuracy():
+    p = profiler.Profiler()
+    p.start()
+    spend_16()
+    p.stop(flush=False)
+    recorder = list(p.recorders)[0]
+    # First index is the stack position, second is the function name
+    time_spent_ns = collections.defaultdict(lambda: collections.defaultdict(lambda: 0))
+    cpu_spent_ns = collections.defaultdict(lambda: collections.defaultdict(lambda: 0))
+    for event in recorder.events[stack.StackSampleEvent]:
+        if event.thread_name == "MainThread":
+            for idx, frame in enumerate(event.frames):
+                time_spent_ns[idx][frame[2]] += event.wall_time_ns
+                cpu_spent_ns[idx][frame[2]] += event.cpu_time_ns
+
+    assert {"spend_3", "spend_1", "spend_cpu_2"}.issubset(set(time_spent_ns[0].keys())), time_spent_ns
+    assert {"spend_4", "spend_7", "spend_16"}.issubset(set(time_spent_ns[1].keys())), time_spent_ns
+
+    assert almost_equal(total_time(time_spent_ns, "spend_3"), 9e9)
+    assert almost_equal(total_time(time_spent_ns, "spend_1"), 2e9)
```

Comment:
```suggestion
    assert almost_equal(total_time(time_spent_ns, "spend_1"), 1e9)
```
?

Reply:
The function is called twice, so its total time is 2 seconds.

Reply:
ah missed that!